WT-16381 Refactor `__disagg_pick_up_checkpoint` function

ablekhman opened the pull request:
- Reinstate `checkpoint=` prefix for the checkpoint info in the turtle file.
- Ensure that turtle file contents follows configuration format.

ablekhman committed:
WT-16381 Refactor `__disagg_pick_up_checkpoint` function
- Reinstate `checkpoint=` prefix for the checkpoint info in the turtle file.
- Ensure that turtle file contents follows configuration format.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -592,64 +530,152 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+    checkpoint_found = timestamp_found = false;
+
+    __wt_config_init(session, &meta_cfg, meta_buf);
+    while ((ret = __wt_config_next(&meta_cfg, &cfg_key, &cfg_value)) == 0) {
```
The shared metadata configs are all separated by newlines, I'm not sure config_next can pass over newlines.

ablekhman replied:
I changed the contents to follow config format. All entries are comma-separated. That's why regular config functions work with it now.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -592,64 +530,152 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+          "Disaggregated checkpoint metadata item \"%.*s\"=\"%.*s\"", (int)cfg_key.len, cfg_key.str,
+          (int)cfg_value.len, cfg_value.str);
+        if (WT_CONFIG_LIT_MATCH("checkpoint", cfg_key)) {
+            if (checkpoint_found)
```
This should feel more like an assert here. If we find a duplicate checkpoint that screams trouble. We can just check if metadata_value is null rather than having checkpoint_found

ablekhman replied:
I'll add more checks once I read path implementation. This function will change again.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -592,64 +530,152 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+                  session, EINVAL, "Duplicate checkpoint entry in disaggregated storage metadata");
+            WT_ERR(__wt_strndup(session, cfg_value.str, cfg_value.len, &metadata_value));
+            checkpoint_found = true;
+        } else if (WT_CONFIG_LIT_MATCH("timestamp", cfg_key)) {
```
Should we check for duplicate timestamp here?

ablekhman replied:
As mentioned above, this function will change after read path.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -592,64 +530,152 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+            else
+                WT_ERR(__wt_txn_parse_timestamp(
+                  session, "checkpoint timestamp", &checkpoint_timestamp, &cfg_value));
+            timestamp_found = true;
```
Do we need timestamp_found? Can we just check for the checkpoint_timestmap being initialized?

ablekhman replied:
Just to be consistent with the checkpoint.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -592,64 +530,152 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+                  session, "checkpoint timestamp", &checkpoint_timestamp, &cfg_value));
+            timestamp_found = true;
+        } else if (WT_CONFIG_LIT_MATCH("key_provider", cfg_key)) {
+            /* Logged above; no additional action required yet. */
```
I would add a FIXME to the read mechanism ticket

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -779,6 +827,98 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+     * Part 2: Get the metadata for other tables from the shared metadata table.
+     */
+
+    WT_ERR(__disagg_process_checkpoint(session, ckpt_meta, internal_session, md_cursor,
```
Does this need this need to pass in the new metadata_value_cfg? I see this function loops through all tables and applies the metadata. Can we call something like `__disagg_apply_checkpoint_meta_all`?

ablekhman replied:
I didn't question the logic of the function. It's quite complex. I just took it as it is. But now I can see that `metadata_value_cfg` can be in param. It is not used after the function call.

ablekhman replied:
`metadata_value_cfg` buffer is owned buy the caller. Yes, it needs to be passed in and out.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -779,6 +827,98 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+    /* Open up a metadata cursor pointing at our table */
+    WT_ERR(__wt_metadata_cursor(internal_session, &md_cursor));
+
+    WT_ERR(__disagg_save_meta(session, internal_session, md_cursor, metadata_value,
```
The open of the internal session -> meta_cursor open -> disagg_save _meta -> disagg_process checkpoint, IMO have a shared responsibility. Which is to all update the metadata information of all tables. 

Can we have a parent function called __disagg_process_metadata which inside does this logic?

Also WDYT about __disagg_apply_shared_meta as a function name?

ablekhman replied:
Both the `internal_session` and the `md_cursor` are reused later in the function. If I refactor this part, then these pointer will have to be out params. It will overcomplicate the code, since the ownership of these objects will be split between calls.  So, I rather keep it as is.

jiechenbo reviewed (commented):
Looking much cleaner! I have some review comments

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -779,6 +827,98 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+     * Part 1: Get the metadata of the shared metadata table and insert it into our metadata table.
+     */
+
+    WT_ERR(__disagg_read_meta(session, ckpt_meta, &buf));
```
How about `__disagg_fetch_shared_meta`?

ablekhman committed:
WT-16381 Refactor `__disagg_pick_up_checkpoint` function
- Reinstate `checkpoint=` prefix for the checkpoint info in the turtle file.
- Ensure that turtle file contents follows configuration format.

ablekhman committed:
WT-16381 Refactor `__disagg_pick_up_checkpoint` function
- Reinstate `checkpoint=` prefix for the checkpoint info in the turtle file.
- Ensure that turtle file contents follows configuration format.

jiechenbo reviewed (approved):
LGTM! Thank you for making it cleaner
